Automattic/jetpack
# CRM: Add React / TypeScript / Jest build configuration

Introduce React / TypeScript / Jest to the Jetpack CRM plugin.
The new configuration is based on how the starter plugin and other Jetpack plugins in the repo structure their code.

I've chosen to add a couple of "not entirely ready for production yet" React components to make the new code manually testable:

1. `AdminPage` which wraps the default Jetpack `AdminPage` component with a CRM specific header and footer.
2. A theoretical example component called `OnboardingWizard` that doesn't do anything else than output hardcoded RNA component.
3. _There's also a `onboarding-wizard-index.jsx` file that only serves the purpose of rendering a new onboarding wizard experience._

Fixes https://github.com/Automattic/zero-bs-crm/issues/2927

## Proposed changes:

**Dummy logic**

* Add a new `Automattic\Jetpack_CRM\Onboarding_Wizard\Boostrap` class hidden behind a feature flag to register and render new React components.
  * Add a new onboarind wizard admin page behind a feature flag so we can continuously start creating it (`jetpack_crm_feature_flag_onboarding_wizard_v2`).
* Add `AdminPage` and `OnboardingWizard` React components.

**JS**

* Add new webpack config export that handles our React processing.
* Add `tsconfig.json` so we start to allow type specific imports etc.
* Add Jest configuration and dependencies so we can write tests.
* Re-introduce `eslintrc.js` so we get linting errors when using incorrect text-domain.
* Ignore `/build/` in `.gitignore` since it's the source for our compiled React.

**PHP**

* Add `Automattic\Jetpack\Assets` package so we can register React scripts "the Jetpack way".
* Add `projects/plugins/crm/src/` as a source for the PHP autoloader in `composer.json`, so future internal packages will be autoloaded by the Jetpack Autoloader.

### Other information:

- [ ] _Have you checked the E2E test CI results, and verified that your changes do not break them?_
- [ ] _Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?_

## Jetpack product discussion



There's no conversations about this.

## Does this pull request change what data or activity we track or use?



No.

## Testing instructions:






* Spin up project locally and install latest dependencies
* Run `jetpack build plugins/crm`
* Add `add_filter( 'jetpack_crm_feature_flag_onboarding_wizard_v2', '__return_true' );` to e.g. `0-sandbox.php` so we register a new page that shows our dummy content.
* Go to `/wp-admin/admin.php?page=zerobscrm-obw`
  * Verify that it looks like the screenshot below.
* Run `pnpm run lint-file projects/plugins/crm/` from the root.
  * Verify that no linting issues pops up.
* Navigate to `projects/plugins/crm`
* Run `pnpm run test`
  * Verify that 2 tests successfully pass:
    * `src/js/components/admin-page/test/index.test.jsx`
    * `src/js/components/onboarding-wizard/test/index.test.jsx`

**Screenshot(s)**

![Screenshot 2023-02-21 at 04 01 56](https://user-images.githubusercontent.com/3846700/220237469-ec95a827-3ee4-4ae1-ab48-3e75d8c7b891.jpg)
